---
title: "ATAC-Seq Data Analysis with Human Data"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r, shorcut, include=FALSE}
## RStudio keyboard shortcut
# Cursor at the beginning of a command line: Ctrl+A
# Cursor at the end of a command line: Ctrl+E
# Clear all the code from your console: Ctrl+L
# Create a pipe operator %>%: Ctrl+Shift+M (Windows) or Cmd+Shift+M (Mac)
# Create an assignment operator <-: Alt+- (Windows) or Option+-(Mac) 
# Knit a document (knitr): Ctrl+Shift+K (Windows) or Cmd+Shift+K (Mac)
# Comment or uncomment current selection: Ctrl+Shift+C (Windows) or Cmd+Shift+C (Mac)
```


# Tutorial
# Align the FASTQ files
## Create a reference genome
```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
mainChromosomes <- paste0("chr", c(1:21,"X","Y","M"))
mainChrSeq <- lapply(mainChromosomes,
                     function(x) BSgenome.Hsapiens.UCSC.hg19[[x]])
names(mainChrSeq) <- mainChromosomes
mainChrSeqSet <- DNAStringSet(mainChrSeq)
Biostrings::writeXStringSet(x = mainChrSeqSet, 
                            filepath = "BSgenome.Hsapiens.UCSC.hg19.mainChrs.fa")
```


## Build the index
```{r}
library(Rsubread)
buildindex(basename = "BSgenome.Hsapiens.UCSC.hg19.mainChrs",
           reference = "BSgenome.Hsapiens.UCSC.hg19.mainChrs.fa",
           indexSplit = TRUE,
           memory = 1000)

# indexSplit: logical indicating whether the index can be split into multiple blocks. The block size is determined by the value of memory. FALSE by default (ie. a single-block index is generated).
# memory: a numeric value specifying the amount of memory (in megabytes) used for storing the index during read mapping. 8000 MB by default. Note that this option is ignored when indexSplit is FALSE.
```


## Align **paired** sequence reads
```{r}
library(ShortRead)
read1 <- readFastq("./data/ENCFF175VOD.fastq.gz")
read2 <- readFastq("./data/ENCFF447BGX.fastq.gz")
writeFastq(read1[1:1000,],"ATACSample_r1.fastq.gz")
writeFastq(read2[1:1000,],"ATACSample_r2.fastq.gz")
# id(read2[1:1000,])
# myRes <- bamQC("~/Downloads/Sorted_ATAC_50K_2.bam")
```


## Align sequence reads using `Rsubread`
```{r}
library(Rsubread)
align(index = "BSgenome.Hsapiens.UCSC.hg19.mainChrs",
      readfile1 = read1, readfile2 = read2,
      output_file = "ATAC_50K_2.bam",
      nthreads = 2, type = 1,
      unique = TRUE, maxFragLength = 2000)
# type: a character string or an integer giving the type of sequencing data. Possible values include rna (or 0; RNA-seq data) and dna (or 1; genomic DNA-seq data such as WGS, WES, ChIP-seq data etc.). Character strings are case insensitive.
```


## Alternative: align sequence reads using `Rbowtie2`
```{r}
library(Rbowtie2)
bowtie2_build(references = "BSgenome.Hsapiens.UCSC.hg19.mainChrs.fa",
              bt2Index = "BSgenome.Hsapiens.UCSC.hg19.mainChrs_bowtie2")
```


## Decompress FASTQ files
```{r}
gunzip("ATAC_Data/ATAC_FQs/SRR891269_1.fastq.gz")
gunzip("ATAC_Data/ATAC_FQs/SRR891269_2.fastq.gz")
```
