---
title: "ATAC-Seq Data Analysis with Human Data - Perform post-alignment processing"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r, shorcut, include=FALSE}
## RStudio keyboard shortcut
# Cursor at the beginning of a command line: Ctrl+A
# Cursor at the end of a command line: Ctrl+E
# Clear all the code from your console: Ctrl+L
# Create a pipe operator %>%: Ctrl+Shift+M (Windows) or Cmd+Shift+M (Mac)
# Create an assignment operator <-: Alt+- (Windows) or Option+-(Mac) 
# Knit a document (knitr): Ctrl+Shift+K (Windows) or Cmd+Shift+K (Mac)
# Comment or uncomment current selection: Ctrl+Shift+C (Windows) or Cmd+Shift+C (Mac)
```


# Tutorial
# Perform ATACseq quality control using the `{ATACseqQC}`
`{ATACseqQC}` includes two useful metrics - PCR Bottleneck Coefficients (PBC1 and PBC2). 

* The `PCRbottleneckCoefficient_1` is calculated as the number of positions in genome with exactly 1 read mapped uniquely compared to the number of positions with at least 1 read. Values less than 0.7 indicate severe bottlenecking, between 0.7 and 0.9 indicate moderate bottlenecking,  greater than 0.9 show no bottlenecking. 

* The `PCRbottleneckCoefficient_2` is calculated as the number of positions in genome with exactly 1 read mapped uniquely compared to the number of positions with exactly 2 reads mapped uniquely. Values less than 1 indicate severe bottlenecking, between 1 and 3 indicate moderate bottlenecking,  greater than 3 show no bottlenecking. 

```{r}
# BiocManager::install("ATACseqQC")
library(ATACseqQC)
ATACQC <- bamQC("./data/Sorted_ATAC_50K_2_ch17.bam")

# Get access to the names of all metrics
names(ATACQC)

# Extract 2 PCR Bottleneck Coefficients
ATACQC$PCRbottleneckCoefficient_1
ATACQC$PCRbottleneckCoefficient_2
```
