---
title: "ATAC-Seq Data Analysis with Human Data - Perform differential anlaysis of peaks"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

# Functional analysis of peaks
```{r}
if (!require("rGREAT", quietly = TRUE)) { BiocManager::install("rGREAT"); library(rGREAT) } else { library(rGREAT) }
great_Job <- submitGreatJob(MacsCalls, species = "hg19")
availableCategories(great_Job)

great_ResultTable <- getEnrichmentTables(great_Job)
names(great_ResultTable)

great_ResultTable[["GO Biological Process"]][1:4, ]
```

## Differential ATACseq - identify changes in open regions between one condition and another condition
Here we will establish a collection of unique peaks that are found in a minimum of 2 samples. These peaks will serve as the basis for evaluating alterations in nucleosome-free ATAC-seq data using DESeq2.
Here is an alternative approach: [DiffBind](https://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf)
```{r}
##----We examine peaks across all our samples and condense them into a unified collection of unique peaks. Subsequently, we generate a matrix indicating whether each peak is present or absent in each sample----
library(GenomicRanges)
library(rtracklayer)
peaks <- dir("./data/ATAC_peaks_count")
peaks_path <- paste0("./data/ATAC_peaks_count/", peaks)
my_peaks <- lapply(peaks_path, ChIPQC:::GetGRanges, simple = T)
my_peaks
# [[1]]
# GRanges object with 19273 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3672288-3672369      *
#       [2]     chr1   3994901-3995006      *
#       [3]     chr1   4142676-4142762      *
#       [4]     chr1   4571603-4571748      *
#       [5]     chr1   4571832-4572068      *
#       ...      ...               ...    ...
#   [19269]     chrY 90767896-90767990      *
#   [19270]     chrY 90773333-90773411      *
#   [19271]     chrY 90784957-90785094      *
#   [19272]     chrY 90801538-90801651      *
#   [19273]     chrY 90813083-90813224      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
# 
# [[2]]
# GRanges object with 19613 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3994858-3994957      *
#       [2]     chr1   4560267-4560343      *
#       [3]     chr1   4571695-4572204      *
#       [4]     chr1   4785443-4785546      *
#       [5]     chr1   4785789-4785866      *
#       ...      ...               ...    ...
#   [19609]     chrY 90761403-90761489      *
#   [19610]     chrY 90767890-90767992      *
#   [19611]     chrY 90784945-90785099      *
#   [19612]     chrY 90808761-90808841      *
#   [19613]     chrY 90813062-90813232      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
# 
# [[3]]
# GRanges object with 29497 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3433991-3434095      *
#       [2]     chr1   3575895-3575972      *
#       [3]     chr1   3670961-3671038      *
#       [4]     chr1   3671534-3671880      *
#       [5]     chr1   3831077-3831155      *
#       ...      ...               ...    ...
#   [29493]     chrY 90773923-90774009      *
#   [29494]     chrY 90784955-90785077      *
#   [29495]     chrY 90801505-90801611      *
#   [29496]     chrY 90807391-90807495      *
#   [29497]     chrY 90813057-90813234      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
# 
# [[4]]
# GRanges object with 36188 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3130336-3130413      *
#       [2]     chr1   3400112-3400250      *
#       [3]     chr1   3433954-3434042      *
#       [4]     chr1   3515020-3515102      *
#       [5]     chr1   3575937-3576078      *
#       ...      ...               ...    ...
#   [36184]     chrY 90784214-90784298      *
#   [36185]     chrY 90784966-90785099      *
#   [36186]     chrY 90785710-90785798      *
#   [36187]     chrY 90801538-90801651      *
#   [36188]     chrY 90829708-90829782      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
# 
# [[5]]
# GRanges object with 43393 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3671695-3671781      *
#       [2]     chr1   4466776-4466928      *
#       [3]     chr1   4602508-4602618      *
#       [4]     chr1   4623435-4623519      *
#       [5]     chr1   4735672-4735746      *
#       ...      ...               ...    ...
#   [43389]     chrY 90784944-90785099      *
#   [43390]     chrY 90793946-90794025      *
#   [43391]     chrY 90801529-90801640      *
#   [43392]     chrY 90805058-90805151      *
#   [43393]     chrY 90812961-90813237      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
# 
# [[6]]
# GRanges object with 27527 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   4471412-4471506      *
#       [2]     chr1   4785622-4785781      *
#       [3]     chr1   4786096-4786271      *
#       [4]     chr1   4807619-4807973      *
#       [5]     chr1   4857557-4857828      *
#       ...      ...               ...    ...
#   [27523]     chrY 90773918-90774009      *
#   [27524]     chrY 90784957-90785099      *
#   [27525]     chrY 90801504-90801596      *
#   [27526]     chrY 90807482-90807562      *
#   [27527]     chrY 90813071-90813231      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths

##----Return a single set of non-redundant peaks----
all_peaks_set <- GRangesList(my_peaks) |> unlist() |> reduce()
class(all_peaks_set)
# [1] "GRanges"
# attr(,"package")
# [1] "GenomicRanges"
# IRanges::reduce: transform all the ranges together as a set to produce a new set of ranges. 

# GRangesList(my_peaks) |> unlist() 
# GRanges object with 175491 ranges and 0 metadata columns:
#            seqnames            ranges strand
#               <Rle>         <IRanges>  <Rle>
#        [1]     chr1   3672288-3672369      *
#        [2]     chr1   3994901-3995006      *
#        [3]     chr1   4142676-4142762      *
#        [4]     chr1   4571603-4571748      *
#        [5]     chr1   4571832-4572068      *
#        ...      ...               ...    ...
#   [175487]     chrY 90773918-90774009      *
#   [175488]     chrY 90784957-90785099      *
#   [175489]     chrY 90801504-90801596      *
#   [175490]     chrY 90807482-90807562      *
#   [175491]     chrY 90813071-90813231      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths

# GRangesList(my_peaks) |> unlist() |> reduce()
# GRanges object with 89769 ranges and 0 metadata columns:
#           seqnames            ranges strand
#              <Rle>         <IRanges>  <Rle>
#       [1]     chr1   3130336-3130413      *
#       [2]     chr1   3400112-3400250      *
#       [3]     chr1   3433954-3434095      *
#       [4]     chr1   3515020-3515102      *
#       [5]     chr1   3575895-3576078      *
#       ...      ...               ...    ...
#   [89765]     chrY 90805058-90805151      *
#   [89766]     chrY 90807391-90807562      *
#   [89767]     chrY 90808761-90808841      *
#   [89768]     chrY 90812961-90813237      *
#   [89769]     chrY 90829708-90829782      *
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths

##----Estimate how often each of these non-redundant peaks occurs in each sample----
overlap <- list()
for (i in 1:length(my_peaks)) {
  overlap[[i]] <- all_peaks_set %over% my_peaks[[i]]
}
# length(my_peaks): 6
# Finding interval overlaps between two "range-based" objects: a query and a subject
# query %over% subject; if query overlaps with subject, return T; otherwise, return F.
overlap_mat <- do.call(cbind, overlap)
dim(overlap_mat) # Return a T/F matrix with 89769 rows and 6 columns
colnames(overlap_mat) <- basename(peaks)
mcols(all_peaks_set) <- overlap_mat # Assign overlap_mat as metadata columns of all_peaks_set

all_peaks_set[1:2, ]

##----We remove peaks found in blacklists and those on ChrM (mitochondrial chromosome) before conducting tests to eliminate the possibility of artifact differential call results----
blacklist <- import.bed("./data/ATAC_blacklists/ENCFF547MET.bed.gz")
eligble_peak <- all_peaks_set[(!all_peaks_set %over% blacklist) & (seqnames(all_peaks_set) != "chrM")]
eligble_peak

##----Determine how many times each of our non-redundant peaks appears in the samples----
elementMetadata(eligble_peak)
# DataFrame with 89654 rows and 6 columns
peak_count <- elementMetadata(eligble_peak) |> as.data.frame() |> rowSums()
eligble_peak_select <- eligble_peak[peak_count >= 2, ]
eligble_peak_select # GRanges object with 36320 ranges and 6 metadata columns

##----We count paired reads landing in peaks----
library(GenomicAlignments)
bam2count <- dir("./data/ATAC_BAM/", full.names = T, pattern = "*.\\.bam$")
# *: Match any sequence of characters.
# \\.: Match a literal dot (.) character. The double backslash (\\) is used to escape the dot because dot has a special meaning in regular expressions.
# bam$: Match the string "bam" at the end of a file name, followed by the end of the line.

my_count <- summarizeOverlaps(features = eligble_peak_select, 
                              reads = bam2count, 
                              singleEnd = F)
# features is A GRanges or a GRangesList object of genomic regions of interest.
colnames(my_count) <- c("hind_brain_1", "hind_brain_2", "kidney_1", "kidney_2", "liver_1", "liver_2")

##----We run DESeq2 for differential ATACseq----
library(DESeq2)
load("./data/my_count.RData")
atac_count <- myCounts
meta_data <- data.frame(Group = factor(c("hind_brain", "hind_brain", "kidney", "kidney", "liver", "liver")),
                        row.names = colnames(atac_count))

atac_dds <- DESeqDataSetFromMatrix(countData = assay(atac_count),
                                   colData = meta_data,
                                   design = ~ Group,
                                   rowRanges = rowRanges(atac_count)) # !
atac_dds <- DESeq(atac_dds)
# estimating size factors
# estimating dispersions
# gene-wise dispersion estimates
# mean-dispersion relationship
# -- note: fitType='parametric', but the dispersion trend was not well captured by the
#   function: y = a/x + b, and a local regression fit was automatically substituted.
#   specify fitType='local' or 'mean' to avoid this message next time.
# final dispersion estimates
# fitting model and testing

kidney_vs_hindbrain <- results(atac_dds, c("Group", "kidney", "hind_brain"), format = "GRanges")
# format	
# character, either "DataFrame", "GRanges", or "GRangesList", whether the results should be printed as a DESeqResults DataFrame, or if the results DataFrame should be attached as metadata columns to the **GRanges** or GRangesList rowRanges of the DESeqDataSet. If the rowRanges is a GRangesList, and GRanges is requested, the **range of each gene will be returned**
kidney_vs_hindbrain_order <- kidney_vs_hindbrain[order(kidney_vs_hindbrain$pvalue)]
kidney_vs_hindbrain_order

# GRanges object with 36320 ranges and 6 metadata columns:
#           seqnames              ranges strand |  baseMean log2FoldChange     lfcSE      stat      pvalue        padj
#              <Rle>           <IRanges>  <Rle> | <numeric>      <numeric> <numeric> <numeric>   <numeric>   <numeric>
#       [1]     chr4   22488379-22488925      * |   96.8199       -3.48476  0.272029 -12.81021 1.43736e-37 5.01798e-33
#       [2]     chr1   24613428-24614006      * |  285.2533        1.46558  0.161832   9.05623 1.35040e-19 2.35719e-15
#       [3]     chrX 143483004-143483128      * | 1130.4129       -1.22535  0.148502  -8.25140 1.56553e-16 1.82181e-12
#       [4]     chr2 150644059-150644430      * |   36.0214       -4.17235  0.521845  -7.99539 1.29168e-15 1.12735e-11
#       [5]    chr15   66239217-66239619      * |   32.3408       -4.44791  0.583590  -7.62164 2.50474e-14 1.74886e-10
#       ...      ...                 ...    ... .       ...            ...       ...       ...         ...         ...
#   [36316]     chrX   93897236-93897401      * |   2.85026              0   2.36122         0           1           1
#   [36317]     chrX 100475652-100475834      * |   2.19112              0   2.56424         0           1          NA
#   [36318]     chrX 153334237-153334437      * |   2.26341              0   2.54823         0           1          NA
#   [36319]     chrX 164197795-164198006      * |   4.02077              0   2.22715         0           1           1
#   [36320]     chrX 169047684-169047857      * |   2.55896              0   2.42940         0           1          NA
#   -------
#   seqinfo: 22 sequences from an unspecified genome; no seqlengths
```
