---
title: "ATAC-Seq Data Analysis with T-Regulatory Cells"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r, shorcut, include=FALSE}
## RStudio keyboard shortcut
# Cursor at the beginning of a command line: Ctrl+A
# Cursor at the end of a command line: Ctrl+E
# Clear all the code from your console: Ctrl+L
# Create a pipe operator %>%: Ctrl+Shift+M (Windows) or Cmd+Shift+M (Mac)
# Create an assignment operator <-: Alt+- (Windows) or Option+-(Mac) 
# Knit a document (knitr): Ctrl+Shift+K (Windows) or Cmd+Shift+K (Mac)
# Comment or uncomment current selection: Ctrl+Shift+C (Windows) or Cmd+Shift+C (Mac)
```

# Exercise
# Preprocess ATAC-seq data
## Read in a random 10000 reads from read 1 and read 2
```{r}
easypackages::libraries("GenomicRanges", "Rsamtools", "rtracklayer", "GenomicAlignments", "ShortRead", "ggplot2")

r1 <- FastqSampler("data/ENCFF175VOD.fastq", n = 10000)
r2 <- FastqSampler("data/ENCFF447BGX.fastq", n = 10000)
# FastqSampler() from {ShortRead}; FastqSampler() draws a subsample from a fastq file. yield() is the method used to extract the sample from the FastqSampler instance

set.seed(888)
p1 <- yield(r1)
set.seed(888)
p2 <- yield(r2)

all.quality <- quality(p1)
# quality(x) from {Biostrings}: The quality of the strings.
quality.boxplot <- as(all.quality, "matrix")
colnames(quality.boxplot) <- paste0("Cycle", 1:ncol(quality.boxplot))
boxplot(quality.boxplot, main = "Read 1")

all.quality <- quality(p2)
quality.boxplot <- as(all.quality, "matrix")
colnames(quality.boxplot) <- paste0("Cycle", 1:ncol(quality.boxplot))
boxplot(quality.boxplot, main = "Read 2")
```

## Create a line plot of Base frequencies across cycles for read 1 and read 2
```{r}
atgc.fr.by.cycle <- alphabetByCycle(sread(p1))
# sread(p1): returns a DNAStringSet object of length 10000
# alphabetByCycle: returns the number of occurrences of each nucleotide (e.g., A, T, G, C) across all cycles for the read 1
class(atgc.fr.by.cycle)
# [1] "matrix" "array" 

atgc.fr.by.cycle.filtered <-  atgc.fr.by.cycle[c("A", "G", "C", "T", "N"), ]

a.fr.by.cycle.df <-
  data.frame(Base = "A",
             Freq = atgc.fr.by.cycle.filtered["A", ],
             Cycle = 1:max(width(sread(p1))))

c.fr.by.cycle.df <-
  data.frame(Base = "C",
             Freq = atgc.fr.by.cycle.filtered["C", ],
             Cycle = 1:max(width(sread(p1))))

t.fr.by.cycle.df <-
  data.frame(Base = "T",
             Freq = atgc.fr.by.cycle.filtered["T", ],
             Cycle = 1:max(width(sread(p1))))

g.fr.by.cycle.df <-
  data.frame(Base = "G",
             Freq = atgc.fr.by.cycle.filtered["G", ],
             Cycle = 1:max(width(sread(p1))))

n.fr.by.cycle.df <-
  data.frame(Base = "N",
             Freq = atgc.fr.by.cycle.filtered["N", ],
             Cycle = 1:max(width(sread(p1))))

read1.df <-
  do.call(rbind, lapply(paste0(c("a", "c", "t", "g", "n"), ".fr.by.cycle.df"), get))

## Not write:  
# myFrameRead1 <- rbind(a.fr.by.cycle.df, c.fr.by.cycle.df, t.fr.by.cycle.df,
#                       g.fr.by.cycle.df, n.fr.by.cycle.df)

read1.df$Read <- "Read1"

atgc.fr.by.cycle <- alphabetByCycle(sread(p2))
atgc.fr.by.cycle.filtered <-  atgc.fr.by.cycle[c("A", "G", "C", "T", "N"), ]

a.fr.by.cycle.df <-
  data.frame(Base = "A",
             Freq = atgc.fr.by.cycle.filtered["A", ],
             Cycle = 1:max(width(sread(p2))))

c.fr.by.cycle.df <-
  data.frame(Base = "C",
             Freq = atgc.fr.by.cycle.filtered["C", ],
             Cycle = 1:max(width(sread(p2))))

t.fr.by.cycle.df <-
  data.frame(Base = "T",
             Freq = atgc.fr.by.cycle.filtered["T", ],
             Cycle = 1:max(width(sread(p2))))

g.fr.by.cycle.df <-
  data.frame(Base = "G",
             Freq = atgc.fr.by.cycle.filtered["G", ],
             Cycle = 1:max(width(sread(p2))))

n.fr.by.cycle.df <-
  data.frame(Base = "N",
             Freq = atgc.fr.by.cycle.filtered["N", ],
             Cycle = 1:max(width(sread(p2))))

read2.df <-
  do.call(rbind, lapply(paste0(c("a", "c", "t", "g", "n"), ".fr.by.cycle.df"), get))

read2.df$Read <- "Read2"

all.df <- rbind(read1.df, read2.df)

ggplot(data = all.df,
       mapping = aes(x = Cycle, y = Freq, colour = Base)) + 
  geom_line() +
  theme_bw() + 
  facet_grid(~ Read)
```

## Align the reads in hg38 genome and create sorted, indexed BAM file
```{r}
# https://genome.ucsc.edu/cgi-bin/hgGateway
library(BSgenome.Hsapiens.UCSC.hg38)

## Extract each chromosome's sequence
chr.name <- paste0("chr", c(1:21, "X", "Y", "M"))
#  [1] "chr1"  "chr2"  "chr3"  "chr4"  "chr5"  "chr6"  "chr7" 
#  [8] "chr8"  "chr9"  "chr10" "chr11" "chr12" "chr13" "chr14"
# [15] "chr15" "chr16" "chr17" "chr18" "chr19" "chr20" "chr21"
# [22] "chrX"  "chrY"  "chrM" 
main.chr.seq <- lapply(chr.name,
                       function(x) BSgenome.Hsapiens.UCSC.hg38[[x]])
names(main.chr.seq) <- chr.name

## Write the DNAStringSet into a FASTA file
main.chr.seq.set <- DNAStringSet(main.chr.seq)
writeXStringSet(main.chr.seq.set,
                "BSgenome.Hsapiens.UCSC.hg38.mainChrs.fa")

library(Rsubread)
buildindex(basename = "BSgenome.Hsapiens.UCSC.hg38.mainChrs",
           reference = "BSgenome.Hsapiens.UCSC.hg38.mainChrs.fa", ## a FASTA or gzipped FASTA file containing the sequences of all chromosomes and contigs
           indexSplit = TRUE,
           memory = 5000)
## An index needs to be built before read alignment can be performed. 
## buildindex() function creates a hash table index for the reference genome, which can then be used by Subread and Subjunc aligners for read alignment.

read1 <- "~/data/ENCFF175VOD.fastq.gz"
read2 <- "~/data/ENCFF447BGX.fastq.gz"
outBAM <- "TReg_ATAC.bam"

align("BSgenome.Hsapiens.UCSC.hg38.mainChrs",
      readfile1 = read1, readfile2 = read2,
      output_file = outBAM,
      nthreads = 2, type = 1,
      unique = TRUE, maxFragLength = 2000)

library(Rsamtools)
sortBam(outBAM, "Sorted_TReg_ATAC")
indexBam("Sorted_TReg_ATAC.bam")
```

## Using the BAM file retrieved from ENCODE, plot the fragment length distribution for reads chromosome 10 using ggplot2
```{r}
library(Rsamtools)
library(ggplot2)
library(GenomicAlignments)
indexBam("data/ENCFF053CGD.bam")

mapped.reads <- idxstatsBam("data/ENCFF053CGD.bam")
my.param <- ScanBamParam(flag = scanBamFlag(isProperPair = TRUE),
                         what = c("qname", "mapq", "isize"),
                         which = GRanges("chr10", IRanges(1, 130694993)))
# 130694993: length of chromosome 10
# https://genome.ucsc.edu/cgi-bin/hgTracks?chromInfoPage=
# qname: reads name; mapq: quality of alignment; isize: insert size of our fragment

pair.reads <- readGAlignmentPairs("data/ENCFF053CGD.bam", param = my.param)

insert.size <- abs(mcols(second(pair.reads))$isize)
# Or insert.size <- elementMetadata(second(pair.reads))$isize |> abs()
fr.insert.size <- table(insert.size)
plot.data <- data.frame(InsertSize = as.numeric(names(fr.insert.size)),
                        Count = as.numeric(fr.insert.size))
frag.len.plot <- ggplot(data = plot.data,
                        mapping = aes(x = InsertSize, y = Count)) + 
  geom_line() + 
  scale_y_continuous(trans = "log2") + 
  theme_minimal()
frag.len.plot
```
