---
title: "ATAC-Seq Data Analysis with Human Data - Perform post-alignment processing"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r, shorcut, include=FALSE}
## RStudio keyboard shortcut
# Cursor at the beginning of a command line: Ctrl+A
# Cursor at the end of a command line: Ctrl+E
# Clear all the code from your console: Ctrl+L
# Create a pipe operator %>%: Ctrl+Shift+M (Windows) or Cmd+Shift+M (Mac)
# Create an assignment operator <-: Alt+- (Windows) or Option+-(Mac) 
# Knit a document (knitr): Ctrl+Shift+K (Windows) or Cmd+Shift+K (Mac)
# Comment or uncomment current selection: Ctrl+Shift+C (Windows) or Cmd+Shift+C (Mac)
```


# Perform post-alignment processing
## [QC] Select properly paired reads using `ScanBamParam()` and `scanBamFlag()`
```{r}
library(GenomicAlignments)
flags <- scanBamFlag(isProperPair = T)
# Here we only select reads paired in alignment within the preset max fragment length 600 bp
# isProperPair = T: represent reads aligning to identical reference sequences and with a specified distance, 600 bp in our case.

# Recap maxFragLength = 600
# align(index = "BSgenome.Hsapiens.UCSC.hg38.mainchr",
#       readfile1 = "./data/SRR6870408_1.fastq.gz", 
#       readfile2 = "./data/SRR6870408_2.fastq.gz",
#       output_file = "./data/ATAC_female_lung.bam",
#       nthreads = 4, type = 1,
#       unique = TRUE, maxFragLength = 600)

chr20_param  <- ScanBamParam(flag = flags,
                             what = c("qname", "mapq", "isize"),
                             which = GRanges("chr20", IRanges(1, 64444167))) 

# https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&chromInfoPage=
# 63025520: length of chromosome 20 (GRCh37/hg19)
# https://genome.ucsc.edu/cgi-bin/hgTracks?hgsid=1795230954_XcEyjdAa9Z968dnsdwUvHaqjwvOg&chromInfoPage=
# 64444167: length of chromosome 20 (GRCh38/hg38)
# qname: reads name
# mapq: quality of alignment
# isize: insert size of our fragment
chr20_param

# class: ScanBamParam
# bamFlag (NA unless specified): isProperPair=TRUE
# bamSimpleCigar: FALSE
# bamReverseComplement: FALSE
# bamTag:  
# bamTagFilter:
# bamWhich: 1 ranges
# bamWhat: qname, mapq, isize
# bamMapqFilter: NA
```

```{r more.chromosomes}
chr1X_param <- ScanBamParam(flag = flags,
                            what = c("qname", "mapq", "isize"),
                            which = IRangesList(chr1=IRanges(1, 248956422),
                                                chrX=IRanges(1, 156040895))) 
```

## Read properly paired reads
```{r}
# Global view of the 2 reads, containing the seqnames, strand, ranges of 2 reads
atac_reads_chr20 <- readGAlignmentPairs(sorted_bam, param = chr20_param)
class(atac_reads_chr20)
# [1] "GAlignmentPairs"
# attr(,"package")
# [1] "GenomicAlignments"
atac_reads_chr20
# GAlignmentPairs object with 686908 pairs, strandMode=1, and 0 metadata columns:
#            seqnames strand   :            ranges  --            ranges
#               <Rle>  <Rle>   :         <IRanges>  --         <IRanges>
#        [1]    chr20      -   :       60100-60196  --       60007-60106
#        [2]    chr20      +   :       60246-60280  --       60248-60288
#        [3]    chr20      +   :       60255-60296  --       60257-60298
#        [4]    chr20      +   :       60260-60296  --       60262-60298
#        [5]    chr20      +   :       60261-60315  --       60263-60323
#        ...      ...    ... ...               ... ...               ...
#   [686904]    chr20      -   : 64332264-64332359  -- 64332180-64332268
#   [686905]    chr20      -   : 64332477-64332521  -- 64332469-64332513
#   [686906]    chr20      +   : 64332967-64333059  -- 64332969-64333061
#   [686907]    chr20      +   : 64333282-64333381  -- 64333312-64333411
#   [686908]    chr20      +   : 64334000-64334042  -- 64334002-64334044
#   -------
#   seqinfo: 25 sequences from an unspecified genome
atac_reads_chr20[1:2, ]
# atac_reads_chr1X <- readGAlignmentPairs(sorted_bam, param = chr1X_param)
# atac_reads_chr1X
# 
# GAlignmentPairs object with 4348926 pairs, strandMode=1, and 0 metadata columns:
#             seqnames strand   :              ranges  --              ranges
#                <Rle>  <Rle>   :           <IRanges>  --           <IRanges>
#         [1]     chr1      +   :          9996-10084  --         10277-10374
#         [2]     chr1      +   :          9997-10087  --         10082-10179
#         [3]     chr1      +   :         10003-10091  --         10123-10221
#         [4]     chr1      +   :         10004-10099  --         10009-10108
#         [5]     chr1      +   :         10004-10104  --         10037-10136
#         ...      ...    ... ...                 ... ...                 ...
#   [4348922]     chrX      -   : 156030710-156030739  -- 156030677-156030713
#   [4348923]     chrX      +   : 156030720-156030793  -- 156030723-156030795
#   [4348924]     chrX      +   : 156030726-156030759  -- 156030728-156030761
#   [4348925]     chrX      -   : 156030736-156030787  -- 156030725-156030785
#   [4348926]     chrX      -   : 156030748-156030783  -- 156030289-156030331
#   -------
#   seqinfo: 25 sequences from an unspecified genome

atac_reads_chr1X <- readGAlignmentPairs(sorted_bam, param = chr1X_param)
atac_reads_chr1X
# GAlignmentPairs object with 4348926 pairs, strandMode=1, and 0 metadata columns:
#             seqnames strand   :              ranges  --              ranges
#                <Rle>  <Rle>   :           <IRanges>  --           <IRanges>
#         [1]     chr1      +   :          9996-10084  --         10277-10374
#         [2]     chr1      +   :          9997-10087  --         10082-10179
#         [3]     chr1      +   :         10003-10091  --         10123-10221
#         [4]     chr1      +   :         10004-10099  --         10009-10108
#         [5]     chr1      +   :         10004-10104  --         10037-10136
#         ...      ...    ... ...                 ... ...                 ...
#   [4348922]     chrX      -   : 156030710-156030739  -- 156030677-156030713
#   [4348923]     chrX      +   : 156030720-156030793  -- 156030723-156030795
#   [4348924]     chrX      +   : 156030726-156030759  -- 156030728-156030761
#   [4348925]     chrX      -   : 156030736-156030787  -- 156030725-156030785
#   [4348926]     chrX      -   : 156030748-156030783  -- 156030289-156030331
#   -------
#   seqinfo: 25 sequences from an unspecified genome

# Gain information on the first or second read, containing the seqnames, strand, cigar, qwidth, start, end, width, qname, mapq, isize
read1 <- first(atacReads)
read2 <- second(atacReads)
read2[1, ]
```


## Retrieve MapQ scores
```{r}
# Do our reads have the high-quality mapping?
read1MapQ <- mcols(read1)$mapq
read2MapQ <- mcols(read2)$mapq
read1MapQ[1:2]
```


## Estimate the MapQ score frequency
```{r}
read1MapQFreqs <- table(read1MapQ)
read2MapQFreqs <- table(read2MapQ)
```


## Plot MapQ scores
```{r}
library(ggplot2)
toPlot <- data.frame(MapQ = c(names(read1MapQFreqs), names(read2MapQFreqs)),
                     Frequency = c(read1MapQFreqs, read2MapQFreqs),
                     Read = c(rep("Read1", length(read1MapQFreqs)), 
                              rep("Read2", length(read2MapQFreqs))))
toPlot$MapQ <- factor(toPlot$MapQ, 
                      levels = unique(sort(as.numeric(toPlot$MapQ))))
ggplot(toPlot, mapping = aes(x = MapQ, y = Frequency, fill = MapQ)) + 
  geom_bar(stat = "identity") + 
  facet_grid(~Read)
```


## Retrieve insert sizes 
```{r}
insertSizes <- elementMetadata(read1)$isize |> abs() # Or use the mcols to retrieve the insert size
head(insertSizes)
```


## Plot insert size
```{r}
library(ggplot2)
fragLenSizes <- table(insertSizes)
fragLenSizes[1:5 ]
toPlot <- data.frame(InsertSize = as.numeric(names(fragLenSizes)),
                     Count = as.numeric(fragLenSizes))
fragLenPlot <- ggplot(toPlot,aes(x = InsertSize, y = Count)) + geom_line()
fragLenPlot + theme_bw() 

# Apply the log2 transformation to the counts to clarify the nucleosome patterning 
fragLenPlot + scale_y_continuous(trans = "log2") + theme_bw()

# Refine the plot
fragLenPlot + scale_y_continuous(trans = "log2") + 
  geom_vline(xintercept = c(180, 247),colour = "red") + 
  geom_vline(xintercept = c(315, 437),colour="darkblue") +
  geom_vline(xintercept = c(100), colour = "darkgreen") + theme_bw()
```


## Subset ATACseq reads by insert sizes
```{r}
atacReads_NucFree <- atacReads[insertSizes < 100, ]
atacReads_MonoNuc <- atacReads[insertSizes > 180 & insertSizes < 240, ]
atacReads_diNuc <- atacReads[insertSizes > 315 & insertSizes < 437, ]
```


## Save the subsetted ATACseq reads as BMA files
```{r}
nucFreeRegionBam <- gsub("\\.bam", "_nucFreeRegions\\.bam", sortedBAM)
monoNucBam <- gsub("\\.bam", "_monoNuc\\.bam", sortedBAM)
diNucBam <- gsub("\\.bam", "_diNuc\\.bam", sortedBAM)

library(rtracklayer)
export(object = atacReads_NucFree, con = nucFreeRegionBam, format = "bam") # Save the object atacReads_NucFree with the filename in `nucFreeRegionBam`
export(object = atacReads_MonoNuc, con = monoNucBam, format = "bam")
export(object = atacReads_diNuc, con = diNucBam, format = "bam")
```


## Reconstruct the fragment using `granges`
```{r}
# Show the GAlignmentPairs object with 1 pair
atacReads[1, ] 

# Show the GRange object with 1 range: collapse read 1 and read2 into one long read
atacFragments <- granges(atacReads)
atacFragments[1, ] 

# Identify the non-redundant fraction of the full-length fragments 

# duplicated() returns a logical vector indicating which elements (rows) are duplicates.
duplicatedFragments <- sum(duplicated(atacFragments))
totalFragments <- length(atacFragments)
duplicateRate <- duplicatedFragments/totalFragments
nonRedundantFraction <- 1 - duplicateRate
nonRedundantFraction # NRF
```


## Construct an open region bigWig
```{r}
openRegionRPMBigWig <- gsub("\\.bam","_openRegionRPM\\.bw",sortedBAM)
myCoverage <- coverage(atacFragments,
                       weight = (10^6/length(atacFragments)))
export.bw(myCoverage,openRegionRPMBigWig)
```
