**Accurately spot cell-free droplets, and provide noise-free scRNAseq quantification**

Fleming, S.J., Chaffin, M.D., Arduini, A. et al. Unsupervised removal of systematic background noise from droplet-based single-cell experiments using CellBender. Nat Methods 20, 1323–1335 (2023). https://doi.org/10.1038/s41592-023-01943-7

<br>

**WGS and RNA-seq based machine learning pipeline for biomarker discovery and predictive analysis**

William DeGroat, Dinesh Mendhe, Atharva Bhusari, Habiba Abdelhalim, Saman Zeeshan, Zeeshan Ahmed, IntelliGenes: A novel machine learning pipeline for biomarker discovery and predictive analysis using multi-genomic profiles, Bioinformatics, 2023;, btad755, https://doi.org/10.1093/bioinformatics/btad755

<br>

**Seperate variations of gene expression enriched in treated single cells from those shared with controls**

Weinberger, E., Lin, C. & Lee, SI. Isolating salient variations of interest in single-cell data with contrastiveVI. Nat Methods 20, 1336–1345 (2023). https://doi.org/10.1038/s41592-023-01955-3

<br>

**Map single-cell fates in regeneration, reprogramming and disease without known directions**

Lange, M., Bergen, V., Klein, M. et al. CellRank for directed single-cell fate mapping. Nat Methods 19, 159–170 (2022). https://doi.org/10.1038/s41592-021-01346-6

<br>

**Expression quantitative trait locus analysis using single cell RNA-Seq data**

Bryois, J., Calini, D., Macnair, W. et al. Cell-type-specific cis-eQTLs in eight human brain cell types identify novel risk genes for psychiatric and neurological disorders. Nat Neurosci 25, 1104–1112 (2022). https://doi.org/10.1038/s41593-022-01128-z

<br>


