## 2023

**Multimodal spatiotemporal phenotyping of human retinal organoid development**

Wahle, P., Brancati, G., Harmel, C. et al. Multimodal spatiotemporal phenotyping of human retinal organoid development. Nat Biotechnol 41, 1765–1775 (2023). https://doi.org/10.1038/s41587-023-01747-2

> "In this study, we generated multiplexed protein maps over a retinal organoid time course and primary adult human retinal tissue...In addition, we generated a single-cell transcriptome and chromatin accessibility timecourse dataset and inferred a gene regulatory network underlying organoid development. We integrated genomic data with spatially segmented nuclei into a multimodal atlas to explore organoid patterning and retinal ganglion cell (RGC) spatial neighborhoods, highlighting pathways involved in RGC cell death and showing that mosaic genetic perturbations in retinal organoids provide insight into cell fate regulation."

<br>

**MultiVI creates a joint representation allowing an analysis of all single cell omics modalities, even for cells for which one or more modalities are missing**

Ashuach, T., Gabitto, M.I., Koodli, R.V. et al. MultiVI: deep generative model for the integration of multimodal data. Nat Methods 20, 1222–1231 (2023). https://doi.org/10.1038/s41592-023-01909-9

<br>

**scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics**

Song, D., Wang, Q., Yan, G. et al. scDesign3 generates realistic in silico data for multimodal single-cell and spatial omics. Nat Biotechnol (2023). https://doi.org/10.1038/s41587-023-01772-1

> "We present a statistical simulator, scDesign3, to generate realistic single-cell and spatial omics data, including various cell states, experimental designs and feature modalities, by learning interpretable parameters from real data. Using a unified probabilistic model for single-cell and spatial omics data, scDesign3 infers biologically meaningful parameters; assesses the goodness-of-fit of inferred cell clusters, trajectories and spatial locations; and generates in silico negative and positive controls for benchmarking computational tools."

<br>

**MISAR-seq for spatially resolved joint profiling of chromatin accessibility and gene expression**

Jiang, F., Zhou, X., Qian, Y. et al. Simultaneous profiling of spatial gene expression and chromatin accessibility during mouse brain development. Nat Methods 20, 1048–1057 (2023). https://doi.org/10.1038/s41592-023-01884-1

> "By applying MISAR-seq to the developing mouse brain, we study tissue organization and spatiotemporal regulatory logics during mouse brain development."

<br>

**GPSA aligns spatial genomics data, allowing for downstream spatially aware analyses**

Jones, A., Townes, F.W., Li, D. et al. Alignment of spatial genomics data using deep Gaussian processes. Nat Methods 20, 1379–1387 (2023). https://doi.org/10.1038/s41592-023-01972-2

> "It remains difficult to precisely align spatial observations across slices, samples, scales, individuals and technologies. Here, we propose a probabilistic model that aligns spatially-resolved samples onto a known or unknown common coordinate system (CCS) with respect to phenotypic readouts (for example, gene expression). Our method, Gaussian Process Spatial Alignment (GPSA), consists of a two-layer Gaussian process: the first layer maps observed samples’ spatial locations onto a CCS, and the second layer maps from the CCS to the observed readouts. Our approach enables complex downstream spatially aware analyses that are impossible or inaccurate with unaligned data, including an analysis of variance, creation of a dense three-dimensional (3D) atlas from sparse two-dimensional (2D) slices or association tests across data modalities."

<br>

**Accurately spot cell-free droplets, and provide noise-free scRNAseq quantification**

Fleming, S.J., Chaffin, M.D., Arduini, A. et al. Unsupervised removal of systematic background noise from droplet-based single-cell experiments using CellBender. Nat Methods 20, 1323–1335 (2023). https://doi.org/10.1038/s41592-023-01943-7

<br>

**WGS and RNA-seq based machine learning pipeline for biomarker discovery and predictive analysis**

William DeGroat, Dinesh Mendhe, Atharva Bhusari, Habiba Abdelhalim, Saman Zeeshan, Zeeshan Ahmed, IntelliGenes: A novel machine learning pipeline for biomarker discovery and predictive analysis using multi-genomic profiles, Bioinformatics, 2023;, btad755, https://doi.org/10.1093/bioinformatics/btad755

<br>

**Identify single-nucleotide variant in scRNA-seq, snRNA-seq, snATAC-seq and scDNA-seq data**

Dou, J., Tan, Y., Kock, K.H. et al. Single-nucleotide variant calling in single-cell sequencing data with Monopogen. Nat Biotechnol (2023). https://doi.org/10.1038/s41587-023-01873-x

> ??? Develop algorithms capable of detecting intronic variants for scRNA-seq, snRNA-seq, snATAC-seq and scDNA-seq data

<br>

**SCENIC+ predicts genomic enhancers along with candidate upstream transcription factors and links these enhancers to candidate target genes**

Bravo González-Blas, C., De Winter, S., Hulselmans, G. et al. SCENIC+: single-cell multiomic inference of enhancers and gene regulatory networks. Nat Methods 20, 1355–1367 (2023). https://doi.org/10.1038/s41592-023-01938-4

> "We exploit SCENIC+ predictions to study conserved TFs, enhancers and GRNs between human and mouse cell types in the cerebral cortex. We use SCENIC+ to study the dynamics of gene regulation along differentiation trajectories and the effect of TF perturbations on cell state."

<br>
 
**Seperate variations of gene expression enriched in treated single cells from those shared with controls**

Weinberger, E., Lin, C. & Lee, SI. Isolating salient variations of interest in single-cell data with contrastiveVI. Nat Methods 20, 1336–1345 (2023). https://doi.org/10.1038/s41592-023-01955-3

<br>

## 2022

**Map single-cell fates in regeneration, reprogramming and disease without known directions**

Lange, M., Bergen, V., Klein, M. et al. CellRank for directed single-cell fate mapping. Nat Methods 19, 159–170 (2022). https://doi.org/10.1038/s41592-021-01346-6

<br>

**Expression quantitative trait locus analysis using single cell RNA-Seq data**

Bryois, J., Calini, D., Macnair, W. et al. Cell-type-specific cis-eQTLs in eight human brain cell types identify novel risk genes for psychiatric and neurological disorders. Nat Neurosci 25, 1104–1112 (2022). https://doi.org/10.1038/s41593-022-01128-z

<br>


