---
title: "Single Cell RNA-Seq Data Analysis with Human Data"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

# Attach the libraries
```{r}
library(Seurat)
# Seurat function manual: https://satijalab.org/seurat/articles/essential_commands.html; https://satijalab.org/seurat/reference/
library(ggplot2)
library(cowplot)
library(purrr)
library(AnnotationHub)
library(readr)
```

# Load the matrix.mtx, features.tsv, and barcodes.tsv files generated from 10x Genomics 
```{r}
sc_data <- Read10X(data.dir = "~/scRNA_analysis/data/filtered_gene_bc_matrices/hg19/")
```

# Create a Seurat object
```{r}
seurat_obj <- CreateSeuratObject(counts = sc_data, project = "scRNAseq", min.cells = 3, min.features = 200)
```

# Scale: load 10x Genomics data from multiple samples, create multiple Seurat objects, and merge all Seurat objects
```{r}
for (folder_name in c("wt_raw_feature_bc_matrix", "kd_raw_feature_bc_matrix", "over_raw_feature_bc_matrix")){
  sc_data <- Read10X(data.dir = paste0("~/scRNA_analysis/data/", folder_name))
  seurat_obj <- CreateSeuratObject(counts = sc_data, 
                                   min.features = 100, 
                                   project = folder_name)
  assign(folder_name, seurat_obj)
}

multi_seurat_obj <- merge(x = wt_raw_feature_bc_matrix, 
                          y = c(kd_raw_feature_bc_matrix, over_raw_feature_bc_matrix),
                          add.cell.id = c("wt", "kd", "over"))
```

# Add meta data
```{r}
meta_data <- multi_seurat_obj@meta.data
meta_data$cell_id <- rownames(meta_data)
meta_data$group <- gsub("^wt_.*", "wt",
  gsub("^kd_.*", "kd",
    gsub("^over_.*", "over", meta_data$cell_id)))
meta_data$batch_id <- gsub("wt", 1,
  gsub("kd", 2,
    gsub("over", 3, meta_data$group)))
multi_seurat_obj@meta.data <- meta_data
```

# Estimate and visualize QC metrics individually and jointly; filter out low-quality cells
```{r}
# QC a single sample
View(seurat_obj@meta.data) # nCount_RNA: number of UMIs per cell; nFeature_RNA: number of genes detected per cell
seurat_obj$log10_gene_per_UMI <- log10(seurat_obj$nFeature_RNA) / log10(seurat_obj$nCount_RNA)
seurat_obj$percent_mt <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
seurat_obj <- subset(x = seurat_obj, subset = nFeature_RNA >= 200 & nCount_RNA >= 500 & log10_gene_per_UMI > 0.8 & percent_mt < 5)

# QC multiple samples
View(multi_seurat_obj@meta.data)
multi_seurat_obj$log10_gene_per_UMI <- log10(multi_seurat_obj$nFeature_RNA) / log10(multi_seurat_obj$nCount_RNA)
multi_seurat_obj$percent_mt <- PercentageFeatureSet(multi_seurat_obj, pattern = "^MT-")

## Cell counts per sample
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(x = group, fill = group)) + 
  geom_bar() +
  ggtitle("Number of unique cellular barcodes detected") + 
  theme_bw()

## UMI counts per cell
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(color = group, x = nCount_RNA, fill = group)) + 
  geom_density(alpha = 0.6) + 
  geom_vline(xintercept = 500)
  scale_x_log10() + 
  ylab("Density") +
  ggtitle("Number of transcripts per cell") + 
  theme_bw()

## Number of genes detected per cell
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(color = group, x = nFeature_RNA, fill = group)) + 
  geom_density(alpha = 0.6) + 
  geom_vline(xintercept = 200)
  scale_x_log10() + 
  ylab("Density") +
  ggtitle("Number of genes detected per cell") + 
  theme_bw()

## Number of genes detected per UMI in a cell
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(x = log10_gene_per_UMI, color = group, fill = group)) +
  geom_density(alpha = 0.6) +
  geom_vline(xintercept = 0.8) + 
  ggtitle("Number of genes detected per UMI in a cell") + 
  theme_bw()

## Mitochondrial percentage
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(color = group, x = percent_mt, fill = group)) + 
  geom_density(alpha = 0.6) + 
  geom_vline(xintercept = 5) + 
  scale_x_log10() + 
  ggtitle("Percentage of transcripts mapping to mitochondrial genes per cell") + 
  theme_bw()

## ! Joint visualization of UMI counts per cell, number of genes detected per cell, and mitochondrial percentage
multi_seurat_obj@meta.data |>
  ggplot(mapping = aes(x = nCount_RNA, y = nFeature_RNA, color = percent_mt)) + 
  geom_point() + 
  geom_vline(xintercept = 500) +
  geom_hline(yintercept = 200) +
  stat_smooth(method = lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  scale_colour_gradient(low = "gray80", high = "black") +
  facet_wrap(~group) + 
  ggtitle("The distributions of joint QC metrics") + 
  theme_bw()

## Filtering using above estimated QC metrics
multi_seurat_obj <- subset(x = multi_seurat_obj, subset = nFeature_RNA >= 200 & nFeature_RNA >= 500 & log10_gene_per_UMI > 0.8 & percent_mt < 5)
multi_seurat_obj_filter <- multi_seurat_obj

## Select genes with a non-zero count appearing in 10 or more cells
count_obj <- GetAssayData(object = multi_seurat_obj_filter, slot = "counts")
count_obj_filter <- count_obj[Matrix::rowSums(count_obj > 0) >= 10, ]
multi_seurat_obj_filter <- CreateSeuratObject(counts = count_obj_filter, meta.data = multi_seurat_obj_filter@meta.data)

## Re-evaluate the most important QC metrics
multi_seurat_obj_filter@meta.data |>
  ggplot(mapping = aes(x = nCount_RNA, y = nFeature_RNA, color = percent_mt)) + 
  geom_point() + 
  geom_vline(xintercept = 500) +
  geom_hline(yintercept = 200) +
  stat_smooth(method = lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  scale_colour_gradient(low = "gray80", high = "black") +
  facet_wrap(~group) + 
  ggtitle("The distributions of joint QC metrics after filtering") + 
  theme_bw()

save(multi_seurat_obj_filter, file = "~/scRNA_analysis/data/derived/multi_seurat_obj_filter.RData")
```

# Assess the sources of variations in the gene expression profiles
## Log normalize the data
```{r}
multi_seurat_obj_lognorm <- NormalizeData(object = multi_seurat_obj_filter, normalization.method = "LogNormalize", scale.factor = 10000)
# scale.factor: to ensure the uniform UMI counts across all cells, here we multiple each UMI count per cell by a cell-specific factor (initial step of normalization); after this initial step, the log transformation will be applied to each UMI count per gene
```

## See if the cell cycle contributes to the major variations in our dataset
```{r}
load("~/scRNA_analysis/data/cycle_marker.rda")

# Assign the cell cycle score to each cell
multi_seurat_obj_lognorm <- CellCycleScoring(object = multi_seurat_obj_lognorm, 
                                             g2m.features = g2m_genes, 
                                             s.features = s_genes)
# s.features: a vector of features associated with S phase
# g2m.features: a vector of features associated with G2M phase

# Check the cell cycle score and cell cycle phase assigned to each cell                             
View(multi_seurat_obj_lognorm@meta.data)  

# Spot the most variable genes
multi_seurat_obj_lognorm <- FindVariableFeatures(multi_seurat_obj_lognorm, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
		     
# Scale the counts for PCA, such that the average expression level of each gene across cells equals 0, and the variance of expression levels of each gene across cells is 1
multi_seurat_obj_lognorm <- ScaleData(multi_seurat_obj_lognorm)

# Run and visualize PCA colored by cell cycle phase
multi_seurat_obj_lognorm <- RunPCA(multi_seurat_obj_lognorm)
DimPlot(object = multi_seurat_obj_lognorm, reduction = "pca", group.by = "Phase", split.by = "Phase")
# Reference: https://satijalab.org/seurat/reference/dimplot
```

## See if the mitochondrial gene expression contributes to the major variations in our dataset
```{r}
percent_mt_q <- quantile(multi_seurat_obj_lognorm@meta.data$percent_mt) 
cut(multi_seurat_obj_lognorm@meta.data$percent_mt_c <- cut(multi_seurat_obj_lognorm@meta.data$percent_mt, breaks = c(-Inf, percent_mt_q[2], percent_mt_q[3], percent_mt_q[4], Inf), labels = c("Low", "Medium", "Medium high", "High"))
DimPlot(multi_seurat_obj_lognorm, reduction = "pca", group.by = "percent_mt_c", split.by = "percent_mt_c")
```

# ! Normalize counts and remove sources of uninteresting variations per condition using sctransform and scaling
```{r}
options(future.globals.maxSize = 4000 * 1024^2)
multi_seurat_obj_split <- SplitObject(multi_seurat_obj_lognorm, split.by = "group")
for (i in 1:length(split_seurat)) {
  multi_seurat_obj_split[[i]] <- SCTransform(multi_seurat_obj_split[[i]], vars.to.regress = c("percent_mt"), vst.flavor = "v2") 
  multi_seurat_obj_split[[i]] <- ScaleData(multi_seurat_obj_split[[i]], vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(multi_seurat_obj_split[[i]]))
  multi_seurat_obj_split[[i]] <- RunPCA(multi_seurat_obj_split[[i]], features = c(s_genes, g2m_genes), verbose = FALSE) # Check if cell cycle impact is regressed out
  DimPlot(multi_seurat_obj_split[[i]], reduction = "pca", group.by = "Phase", split.by = "Phase")
}

multi_seurat_obj_split$wt@assays
multi_seurat_obj_split$kd@assays
multi_seurat_obj_split$over@assays

saveRDS(multi_seurat_obj_split, file = "~/scRNA_analysis/data/derived/multi_seurat_obj_split.rds")
```

# Integrate cells across conditions
## See if there is any condition-specific clustering of the cells
```{r}
multi_seurat_obj_lognorm <- RunUMAP(multi_seurat_obj_lognorm, dims = 1:40, reduction = "pca")
DimPlot(multi_seurat_obj_lognorm)
```

## Perform integration using canonical correlation analysis
```{r}
feature_integrate <- SelectIntegrationFeatures(object.list = multi_seurat_obj_split, nfeatures = 3000) # Choose the most variable features used in integration
multi_seurat_obj_split <- PrepSCTIntegration(object.list = multi_seurat_obj_split, anchor.features = feature_integrate)
anchor_integrate <- FindIntegrationAnchors(object.list = multi_seurat_obj_split, normalization.method = "SCT", anchor.features = feature_integrate)
multi_seurat_obj_integrate <- IntegrateData(anchorset = anchor_integrate, normalization.method = "SCT")
```

## See if integration works well such that each cluster is made up of cells from both conditions
```{r}
multi_seurat_obj_integrate <- RunPCA(object = multi_seurat_obj_integrate)
PCAPlot(multi_seurat_obj_integrate, split.by = "group")  
set.seed(98966)
multi_seurat_obj_integrate <- RunUMAP(multi_seurat_obj_integrate, dims = 1:40, reduction = "pca")
DimPlot(multi_seurat_obj_integrate) 
DimPlot(multi_seurat_obj_integrate, split.by = "group") 

saveRDS(multi_seurat_obj_integrate, "~/scRNA_analysis/data/derived/multi_seurat_obj_integrate.rds")
```

## ! Integrate cells across multiple confounding factors (e.g., experimental conditions donor, tissue/location, species, protocol) using `Harmony`
```{r}
BiocManager::install("harmony", version = "3.8")
library(harmony)
multi_seurat_obj_split <- readRDS("~/scRNA_analysis/data/derived/multi_seurat_obj_split.rds")
feature_integrate <- SelectIntegrationFeatures(object.list = multi_seurat_obj_split, nfeatures = 3000)
multi_seurat_obj_merge <- merge(x = multi_seurat_obj_split[[1]], y = multi_seurat_obj_split[2:length(multi_seurat_obj_split)],
		                merge.data = TRUE)
DefaultAssay(multi_seurat_obj_merge) <- "SCT"
VariableFeatures(multi_seurat_obj_merge) <- feature_integrate
multi_seurat_obj_merge <- RunPCA(multi_seurat_obj_merge, assay = "SCT", npcs = 50)
multi_seurat_obj_harmony <- RunHarmony(multi_seurat_obj_merge, group.by.vars = c("group", "batch_id"), reduction = "pca", assay.use = "SCT", reduction.save = "harmony")

# See if Harmony integration is effective in aligning cells across multiple factors
multi_seurat_obj_harmony <- RunUMAP(multi_seurat_obj_harmony, reduction = "harmony", assay = "SCT", dims = 1:40)
DimPlot(multi_seurat_obj_harmony) 
```

# Old fashioned normalization on counts in a single sample
## Normalize the data
```{r}
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 10000)
```

## Find variable features
```{r}
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(seurat_obj), 10)
```

## Scale the data
```{r}
seurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))
```

# Cluster the cells
## Determine the number of principal components used in the cell clustering; these principal components are assumed to capture the major source of variations driving the cell types
```{r}
ElbowPlot(object = multi_seurat_obj_harmony, ndims = 40)
percent_variation_per_pc <- multi_seurat_obj_harmony[["pca"]]@stdev / sum(multi_seurat_obj_harmony[["pca"]]@stdev) * 100
cumulative_percent_variation_per_pc <- cumsum(percent_variation_per_pc)
pc_number_1 <- which(percent_variation_per_pc < 5 & cumulative_percent_variation_per_pc > 90)[1]
pc_number_2 <- sort(which((percent_variation_per_pc[1:length(percent_variation_per_pc) - 1] - percent_variation_per_pc[2:length(percent_variation_per_pc)]) > 0.1), decreasing = T)[1] + 1
min(pc_number_1, pc_number_2)

elbow_data <- data.frame(pct_pc = percent_variation_per_pc, cum_pc = cumulative_percent_variation_per_pc, rank_pc = 1:length(percent_variation_per_pc))
elbow_data |>
  ggplot(mapping = aes(cum_pc, pct_pc, label = rank_pc, color = rank_pc > min(pc_number_1, pc_number_2))) + 
  geom_text() + 
  geom_hline(yintercept = min(pct_pc[pct_pc > 5]), color = "red") +
  geom_vline(xintercept = 90, color = "red") + 
  labs(x = "Cumulative percent variation per pc", y = "Percent variation per pc", title = "Elbow plot with additional quantitative info") + 
  theme_bw()
```

## Start clustering under different resolutions
```{r}
# seurat_obj <- FindNeighbors(seurat_obj, dims = 1:40)
# seurat_obj <- FindClusters(seurat_obj, resolution = seq(0.2, 1.2, 0.2))
# resolution: above (below) 1.0 -> get a larger (smaller) number of communities

# If integrate cells using canonical correlation analysis
multi_seurat_obj_integrate <- FindNeighbors(multi_seurat_obj_integrate, dims = 1:40)
multi_seurat_obj_integrate <- FindClusters(multi_seurat_obj_integrate, resolution = seq(0.2, 1.2, 0.2))

# If integrate cells using harmony integration
multi_seurat_obj_harmony <- FindNeighbors(object = multi_seurat_obj_harmony, reduction = "harmony")
multi_seurat_obj_harmony <- FindClusters(multi_seurat_obj_harmony, resolution = seq(0.2, 1.2, 0.2))
```

## Plot the cell clustering
```{r}
multi_seurat_obj_harmony@meta.data |> View()
Idents(object = multi_seurat_obj_harmony) <- "integrated_snn_res.0.6" # TD
# Recap: multi_seurat_obj_harmony <- RunUMAP(multi_seurat_obj_harmony, reduction = "harmony", assay = "SCT", dims = 1:40)
DimPlot(multi_seurat_obj_harmony, reduction = "umap", label = TRUE, label.size = 7)

for (i in seq(0.2, 1.2, 0.2)) {
  Idents(object = multi_seurat_obj_harmony) <- sprintf("integrated_snn_res.%s", i)
  DimPlot(multi_seurat_obj_harmony, reduction = "umap", label = TRUE, label.size = 7)
}
```

# Examine the QC metrics of cell clusters
## Distribution of cells per cluster in each condition/sample
```{r}
count_cell_cluster_condition <- FetchData(multi_seurat_obj_harmony, vars = c("ident", "orig.ident")) |> 
  dplyr::count(ident, orig.ident) |>
  tidyr::spread(key = ident, value = n)

View(count_cell_cluster_condition)
```

## Plot cell clustering per condition/sample
```{r}
DimPlot(multi_seurat_obj_harmony, label = TRUE, split.by = "group")  + NoLegend()
```

## Plot cell clustering by the different cell cycle phases
```{r}
DimPlot(multi_seurat_obj_harmony, label = TRUE, split.by = "Phase")  + NoLegend()
```

## Plot cell clustering by additional metrics
```{r}
# Reference: https://satijalab.org/seurat/reference/featureplot
FeaturePlot(object = multi_seurat_obj_harmony, 
            reduction = "umap", 
            features = c("nCount_RNA", "nFeature_RNA", "percent_mt", "S.Score", "G2M.Score"),
            pt.size = 0.7, order = TRUE, min.cutoff = "q10", label = TRUE)
```

## Plot cell clustering by the different principal components
```{r}
cols <- c(paste0("PC_", 1:10), "ident", "umap_1", "umap_2")
pc_data <- FetchData(object = multi_seurat_obj_harmony, vars = cols)

umap_cluster_tag <- FetchData(multi_seurat_obj_harmony, vars = cols[-1:-10]) |>
  group_by(ident) |>
  summarise(x = mean(umap_1), y = mean(umap_2))

pc_vis_fun <- function(pc_id) {
  pc_data |> 
    ggplot(mapping = aes(umap_1, umap_2)) +
      geom_point(aes_string(color = pc_id), alpha = 0.8) +
      scale_color_gradient(guide = FALSE, low = "grey80", high = "purple")  +
      geom_text(data = umap_cluster_tag, mapping = aes(label = ident, x, y)) +
      labs(title = pc_id)
}
  
map(paste0("PC_", 1:10), pc_vis_fun) |> 
  plot_grid(plotlist = .)

print(multi_seurat_obj_harmony[["pca"]], dims = 1:10, nfeatures = 6)
```

# Determine cell identities of each cluster
## Exploration stage
```{r}
DimPlot(object = multi_seurat_obj_harmony, reduction = "umap", label = TRUE) + NoLegend()
DefaultAssay(multi_seurat_obj_harmony) <- "RNA"
multi_seurat_obj_harmony <- NormalizeData(object = multi_seurat_obj_harmony, normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE)

feature_plot_fun <- function(known_gene_marker) {
  FeaturePlot(multi_seurat_obj_harmony, 
              reduction = "umap", 
              features = known_gene_marker, 
              order = TRUE,
              min.cutoff = 'q10', 
              label = TRUE)
}
known_gene_marker_set <- list(c("FCGR3A", "MS4A7"), c("CD79A", "MS4A1"), "CD3D", c("CD3D", "IL7R", "CCR7"), c("CD3D", "CD8A"), 
                              c("GNLY", "NKG7"), "PPBP", c("HBB", "HBA2"), c("CD14", "LYZ"), c("MARCO", "ITGAM", "ADGRE1"), 
                              c("FCER1A", "CST3"), c("IL3RA", "GZMB", "SERPINF1", "ITM2C"))
# Gene marker resources: https://www.10xgenomics.com/analysis-guides/web-resources-for-cell-type-annotation

map(known_gene_marker_set, feature_plot_fun)
```

## Marker identification stage
### Identify unknown clusters and increase confidence in hypothesized cell types for a single sample group/condition - `FindAllMarkers()`
```{r}
derived_gene_marker <- FindAllMarkers(object = multi_seurat_obj_harmony, only.pos = TRUE, logfc.threshold = 0.25)   
```

### Identify clusters that are conserved across conditions for multiple sample groups/conditions - `FindConservedMarkers()`
```{r}
# Find conserved markers for one cluster
cluster0_derived_marker <- FindConservedMarkers(object = multi_seurat_obj_harmony, ident.1 = 0, grouping.var = "group", only.pos = TRUE, logfc.threshold = 0.25)
# ident.2: a second identity class for comparison. If NULL (default) - use all other cells for comparison.
# logfc.threshold = 0.25 -> ~20% increase in gene expression

anno_con <- AnnotationHub()
anno_db <- query(ah, pattern = c("Homo sapiens", "EnsDb"), ignore.case = TRUE)
latest_id <- tail(rownames(mcols(anno_db)), n = 1)
ensemble_db <- anno_con[[latest_id]]
anno_data <- genes(ensemble_db, return.type = "data.frame")[, c("gene_id", "gene_name", "seq_name", "gene_biotype", "description")]
cluster0_derived_marker_anno <- cluster0_derived_marker |> 
  rownames_to_column(var = "gene") |>
  left_join(y = unique(anno_data[, c("gene_name", "description")]),
            by = c("gene" = "gene_name"))

cluster0_derived_marker_anno |> View()

# Find conserved markers for all clusters
cluster_derived_marker_anno_fun <- function(cluster_num) {
  # This function estimates the conserved markers for any cluster
  FindConservedMarkers(multi_seurat_obj_harmony, ident.1 = cluster_num, grouping.var = "group", only.pos = TRUE) |>
    rownames_to_column(var = "gene") |>
    left_join(y = unique(anno_data[, c("gene_name", "description")]),
              by = c("gene" = "gene_name")) |>
    cbind(cluster_id = cluster_num, .)
}
all_cluster_derived_marker_anno <- map_dfr(1:12, cluster_derived_marker_anno_fun)

# Determine top 20 markers per cluster
all_cluster_derived_marker_anno_top20 <- all_cluster_derived_marker_anno |>
  mutate(avg_log2FC = (kd_avg_log2FC + over_avg_log2FC + wt_avg_log2FC) / 3) |>
  group_by(cluster_id) |>
  top_n(n = 20, 
        wt = avg_log2FC) # wt: the variable to use for ordering

all_cluster_derived_marker_anno_top20 |> View()

# Plot top derived marker genes of interest
FeaturePlot(object = multi_seurat_obj_harmony, features = c("CCR7", "CD3D", "LTB"),
            order = TRUE, min.cutoff = 'q10', label = TRUE, repel = TRUE)

VlnPlot(object = multi_seurat_obj_harmony, features = c("CCR7", "CD3D", "LTB"))
```

### Identify subpopulations of the same cell types - `FindMarkers()`
```{r}
cluster_derived_marker_anno_omnipotent <- function(cluster_num_1, cluster_num_2, marker_type = c("all", "conserved", "subpopulation")) {
  # This function estimates the all/conserved/subpopulation markers for any cluster
  marker_type <- match.arg(marker_type, c("all", "conserved", "subpopulation"))
  if (marker_type == "all") {
    FindAllMarkers(multi_seurat_obj_harmony, only.pos = TRUE, logfc.threshold = 0.25)   
  } else if (marker_type == "conserved") { 
      FindConservedMarkers(multi_seurat_obj_harmony, ident.1 = cluster_num_1, grouping.var = "group", only.pos = TRUE) |>
        rownames_to_column(var = "gene") |>
        left_join(y = unique(anno_data[, c("gene_name", "description")]),
                  by = c("gene" = "gene_name")) |>
        cbind(cluster_id = cluster_num_1, .)
  } else {
    FindMarkers(multi_seurat_obj_harmony, ident.1 = cluster_num_1, ident.2 = cluster_num_2) |>
      rownames_to_column(var = "gene") |>
      left_join(y = unique(anno_data[, c("gene_name", "description")]),
                by = c("gene" = "gene_name"))
  }
}

cluster0_derived_marker_anno <- cluster_derived_marker_anno_omnipotent(cluster_num_1 = 0, cluster_num_2 = c(5, 6, 8), marker_type = "subpopulation")

cluster0_derived_marker_anno |>
  dplyr::arrange(p_val_adj) 
```

### Label identity of each cluster
```{r}
multi_seurat_obj_harmony <- RenameIdents(object = multi_seurat_obj_harmony, 
                                         "0" = "CD4+ Naive cells",
                                  	 "1" = "CD4+ Th1 cells",
                                 	 "2" = "CD4+ Th17 cells",
                                  	 "3" = "CD4+ Treg cells",
	                                 "4" = "CD8+ Naive cells",
	                                 "5" = "CD8+ CM cells",
	                                 "6" = "CD8+ EM cells",
	                                 "7" = "CD8+ TEMRA cells",
	                                 "8" = "NK cells",
	                                 "9" = "B cells",
	                                 "10" = "Activated T cells",
	                                 "11" = "MAIT",
	                                 "12" = "CD14+ monocytes",
	                                 "13" = "Myeloid dendritic cells",
	                                 "14" = "Plasmacytoid dendritic cells",
	                                 "15" = "Unknown")

DimPlot(object = multi_seurat_obj_harmony, reduction = "umap", label = TRUE, label.size = 3.5, repel = TRUE)

# Remove unknown cells
multi_seurat_obj_harmony_2 <- subset(multi_seurat_obj_harmony, idents = "Unknown", invert = TRUE)

DimPlot(object = multi_seurat_obj_harmony_2, reduction = "umap", label = TRUE, label.size = 3.5, repel = TRUE)

# write_rds() is a wrapper of saveRDS(), which does not compress the file by default
write_rds(multi_seurat_obj_harmony_2, file = "multi_seurat_obj_harmony_2.rds")
```
