---
title: "Single Cell RNA-Seq Data Analysis with Human Data"
author: "Anni Liu"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r}
library(Seurat)
```

# Load the matrix.mtx, features.tsv, and barcodes.tsv files generated from 10x Genomics 
```{r}
sc_data <- Read10X(data.dir = "~/scRNA_analysis/data/filtered_gene_bc_matrices/hg19/")
```

# Create a Seurat object
```{r}
seurat_obj <- CreateSeuratObject(counts = sc_data, project = "scRNAseq", min.cells = 3, min.features = 200)
```

# Scale: load 10x Genomics data from multiple samples, create multiple Seurat objects, and merge all Seurat objects
```{r}
for (folder_name in c("wt_raw_feature_bc_matrix", "kd_raw_feature_bc_matrix", "over_raw_feature_bc_matrix")){
  sc_data <- Read10X(data.dir = paste0("~/scRNA_analysis/data/", folder_name))
  seurat_obj <- CreateSeuratObject(counts = sc_data, 
                                   min.features = 100, 
                                   project = folder_name)
  assign(folder_name, seurat_obj)
}

multi_seurat_obj <- merge(x = wt_raw_feature_bc_matrix, 
                          y = c(kd_raw_feature_bc_matrix, over_raw_feature_bc_matrix),
                          add.cell.id = c("wt", "kd", "over"))
```

# Estimate QC metrics and filter out low-quality cells
```{r}
# QC a single sample
View(seurat_obj@meta.data) # nCount_RNA: number of UMIs per cell; nFeature_RNA: number of genes detected per cell
seurat_obj$log10_gene_per_UMI <- log10(seurat_obj$nFeature_RNA) / log10(seurat_obj$nCount_RNA)
seurat_obj$percent_mt <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
seurat_obj <- subset(x = seurat_obj, subset = nFeature_RNA >= 200 & nFeature_RNA >= 500 & log10_gene_per_UMI > 0.8 & percent_mt < 5)

# QC multiple samples
View(multi_seurat_obj@meta.data)
multi_seurat_obj$log10_gene_per_UMI <- log10(multi_seurat_obj$nFeature_RNA) / log10(multi_seurat_obj$nCount_RNA)
multi_seurat_obj$percent_mt <- PercentageFeatureSet(multi_seurat_obj, pattern = "^MT-")
multi_seurat_obj <- subset(x = multi_seurat_obj, subset = nFeature_RNA >= 200 & nFeature_RNA >= 500 & log10_gene_per_UMI > 0.8 & percent_mt < 5)
```

# Normalize the data
```{r}
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 10000)
```

# Find variable features
```{r}
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(seurat_obj), 10)
```

# Scale the data
```{r}
seurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))
```

# Perform linear dimensional reduction
```{r}
seurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(object = seurat_obj))
```

# Cluster the cells
```{r}
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)
seurat_obj <- FindClusters(seurat_obj, resolution = 0.5)
```

# Run non-linear dimensional reduction (t-SNE or UMAP)
```{r}
seurat_obj <- RunUMAP(seurat_obj, dims = 1:10)
```

# Save results
```{r}
saveRDS(seurat_obj, file = "seurat_obj.rds")
```
